byron/ledger/executable-spec/test/Ledger/Relation/Properties.hs

```diff
@@ -0,0 +1,223 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ledger.Relation.Properties
+    (testRelation)
+  where
+
+import           Data.Bimap          (Bimap)
+import qualified Data.Bimap          as Bimap
+import           Data.Map.Strict     (Map)
+import qualified Data.Map.Strict     as Map
+import           Data.Set            (Set)
+import qualified Data.Set            as Set
+
+import           Hedgehog            (Gen, MonadTest, Property, PropertyT,
+                                      classify, forAll, property, withTests,
+                                      (===))
+
+import qualified Hedgehog.Gen        as Gen
+import qualified Hedgehog.Range      as Range
+
+import           Test.Tasty.Hedgehog
+
+import           Ledger.Core         hiding ((<|))
+
+import           Test.Tasty          (TestTree, testGroup)
+
+--------------------------------------------------------------------------------
+-- Properties on Relations
+--------------------------------------------------------------------------------
+
+-- | (dom r) ∩ s == dom (s ◁ r)
+--
+--  By restricting the domain of 'Relation r' to 's', the new 'Domain r'
+--  is an intersection of the original 'Domain r' with 's'
+propDomainRestrictionAndIntersection
+  :: (MonadTest m, Relation r, Ord (Domain r))
+  => Domain r -> Set (Domain r) -> r -> m ()
+propDomainRestrictionAndIntersection x s r
+  = Set.member x (dom r `Set.intersection` s)
+    ===
+    Set.member x (dom (s ◁ r))
+
+-- | (dom r ∩ s) ◁ r == s ◁ r
+--
+--  Restricting the domain of 'Relation r' to 's'
+--  is the same as restricting  the domain to 'dom r ∩ s'.
+propDomainRestrictionAndIntersectionB
+  :: (MonadTest m, Relation r, Ord (Domain r), Eq (Range r), Show (Range r))
+  => (Domain r -> r -> Maybe (Range r))
+  -> Domain r
+  -> Set (Domain r)
+  -> r
+  -> m ()
+propDomainRestrictionAndIntersectionB find x s r
+  = find x ((dom r `Set.intersection` s) ◁ r)
```

Comment:
Why do we use `find` and not set equality?

Reply:
It was a remnant of my early draft ;) I've switched to Equality